ethereum-consensus/src/altair/block_processing.rs

```diff
@@ -159,6 +156,103 @@ pub fn process_attestation<
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const SYNC_COMMITTEE_SIZE: usize,
+>(
+    state: &mut BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        SYNC_COMMITTEE_SIZE,
+    >,
+    public_key: BlsPublicKey,
+    withdrawal_credentials: Bytes32,
+    amount: u64,
+    context: &Context,
+) {
+    state.validators.push(get_validator_from_deposit(
+        public_key,
+        withdrawal_credentials,
+        amount,
+        context,
+    ));
+    state.balances.push(amount);
+    state.previous_epoch_participation.push(ParticipationFlags::default());
+    state.current_epoch_participation.push(ParticipationFlags::default());
+    state.inactivity_scores.push(0)
+}
+
+pub fn apply_deposit<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const SYNC_COMMITTEE_SIZE: usize,
+>(
+    state: &mut BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        SYNC_COMMITTEE_SIZE,
+    >,
+    public_key: &BlsPublicKey,
+    withdrawal_credentials: &Bytes32,
+    amount: u64,
+    signature: &BlsSignature,
+    context: &Context,
+) -> Result<()> {
+    let index = state
+        .validators
+        .iter()
+        .enumerate()
+        .find(|(_, v)| v.public_key == *public_key)
+        .map(|(i, _)| i);
+    if let Some(index) = index {
+        increase_balance(state, index, amount);
+        return Ok(());
+    }
+
+    let mut deposit_message = DepositMessage {
+        public_key: public_key.clone(),
+        withdrawal_credentials: withdrawal_credentials.clone(),
+        amount,
+    };
+    let domain = compute_domain(DomainType::Deposit, None, None, context)?;
+    let signing_root = compute_signing_root(&mut deposit_message, domain)?;
+    if verify_signature(public_key, signing_root.as_ref(), signature).is_err() {
+        // NOTE: explicitly return with no error and also no further mutations to `state`
+        return Ok(());
+    }
+
+    add_validator_to_registry(
+        state,
+        public_key.clone(),
+        withdrawal_credentials.clone(),
+        amount,
+        context,
+    );
+
+    Ok(())
+}
+
 pub fn process_deposit<
```

Comment:
I think we should also be able to delete this `process_deposit` definition and let the `spec-gen` handle it